File: haskell/src/Tree/BinarySearchTree.hs
module BinarySearchTree where

import Data.Boolean

data Tree a = EmptyTree | Node { value :: a, left :: Tree a, right :: Tree a, count :: Int } deriving (Show, Read, Eq)

singleton :: (Ord a) => a -> Tree a
singleton x = (Node x EmptyTree EmptyTree 1)

insert :: (Ord a) => Tree a -> a -> Tree a
insert EmptyTree x = singleton x
insert t x
    | x < value t = if left t == EmptyTree
                    then (Node (value t) (singleton x) (right t) (count t))
                    else (Node (value t) (insert (left t) x) (right t) (count t))
    | x > value t = if right t == EmptyTree
                    then (Node (value t) (left t) (singleton x) (count t))
                    else (Node (value t) (left t) (insert (right t) x) (count t))
    | otherwise = (Node (value t) (left t) (right t) (count t + 1))

delete :: (Ord a) => Tree a -> a -> Tree a
delete t x
    | x < value t          = if left t /= EmptyTree
                            then (Node (value t) (delete (left t) x) (right t) (count t))
                            else t
    | x > value t          = if right t /= EmptyTree
                            then (Node (value t) (left t) (delete (right t) x) (count t))
                            else t
    | count t > 1          = (Node (value t) (left t) (right t) (count t - 1))
    | left t /= EmptyTree  = let sub = rightmost $ left t -- replace with the rightmost node of the left subtree
                                 newLeft = hardDelete (left t) (value sub)
                             in (Node (value sub) newLeft (right t) (count sub))
    | right t /= EmptyTree = let sub = leftmost $ right t -- replace with the leftmost node of the right subtree
                                 newRight = hardDelete (right t) (count sub)
                             in (Node (value sub) (left t) newRight (count sub))
    | otherwise            = EmptyTree -- delete this node entirely

-- delete a node regardless of its count, for internal use
hardDelete :: (Ord a) => Tree a -> a -> Tree a
hardDelete t x
    | x < value t          = if left t /= EmptyTree
                             then (Node (value t) (hardDelete (left t) x) (right t) (count t))
                             else t
    | x > value t          = if right t /= EmptyTree
                             then (Node (value t) (left t) (hardDelete (right t) x) (count t))
                             else t
    | left t /= EmptyTree  = let sub = rightmost $ left t -- replace with the rightmost node of the left subtree
                                 newLeft = hardDelete (left t) (value sub)
                             in (Node (value sub) newLeft (right t) (count sub))
    | right t /= EmptyTree = let sub = leftmost $ right t -- replace with the leftmost node of the right subtree
                                 newRight = hardDelete (right t) (count sub)
                             in (Node (value sub) (left t) newRight (count sub))
    | otherwise            = EmptyTree -- delete this node entirely

leaf :: Tree a -> Bool
leaf t = and [left t == EmptyTree, right t == EmptyTree]

singleChild :: Tree a -> bool
singleChild t = if left t == EmptyTree -- hard coded xor since there's no prelude function for it
                then right t /= EmptyTree
                else right t == EmptyTree

-- returns the rightmost node of the tree
rightmost :: Tree a -> Tree a
rightmost t
    | right t == EmptyTree = t
    | otherwise            = rightmost $ right t

-- returns the leftmost node of the tree
leftmost :: Tree a -> Tree a
leftmost t
    | left t == EmptyTree = t
    | otherwise           = leftmost $ left t

File: go/src/tree/binary_search_tree.go
package tree

type BinarySearchTree struct {
  val int
  count int
  left *BinarySearchTree
  right *BinarySearchTree
  parent *BinarySearchTree
}

// func NewBST(val int, parent *BinarySearchTree) *BinarySearchTree {
//   t := new(BinarySearchTree)
//   t.val = val
//   t.parent = parent
//   t.count = 1
//   return t
// }

// func (t *BinarySearchTree) Insert(val int) {
//   newNode := NewBST(val, t)

//   if val == t.val {
//     t.count++
//   } else if val < t.val {
//     if t.left != nil {
//       t.left.Insert(val)
//     } else {
//       t.left = newNode
//     }
//   } else {
//     if t.right != nil {
//       t.right.Insert(val)
//     } else {
//       t.right = newNode
//     }
//   }
// }

// // remove the node with the given value
// func (t *BinarySearchTree) Remove(val int) {
//   // allows for nulling the reference to the child if
//   // it's the thing to be deleted
//   if t.removeChildNode(val) { return }

//   if val < t.val && t.left != nil {
//     t.left.Remove(val)
//   } else if val > t.val && t.right != nil {
//     t.right.Remove(val)
//   } else if t.count > 1 {
//     t.count--
//   } else {
//     // actually remove the node
//     // if it has a left node, replace it with the rightmost descendant of the left node
    
// 	if t.left != nil || t.right != nil {
// 		replacement = t.left != nil ? t.left.max() : t.right.min()

// 		t.hardDelete(replacement)
// 		t.val = replacement.val
// 		t.count = replacement.count 
// 	} else {
// 		// the node's a leaf, just remove it directly
// 		if t.parent.val > t.val {
// 			t.parent.right = nil
// 		} else {
// 			t.parent.left = nil
// 		}
// 	}
//   }
// }

// func (t *BinarySearchTree) hardDelete(val int) {
// 	if t.val
// }

// func (t *BinarySearchTree) max() *BinarySearchTree {
// 	if t.right != nil {
// 		return t.right.max()
// 	} else {
// 		return t
// 	}
// }

// func (t *BinarySearchTree) min() *BinarySearchTree {
// 	if t.left != nil {
// 		return t.left.max()
// 	} else {
// 		return t
// 	}
// }
